File: Source/Loaders/iOS-Keyboard/KeyKeyEngine/Sources/KeyKeyEngine/TouchSmartHostTextState.swift
/// Tracks the editable tail after it has been inserted as ordinary host text.
/// Text evicted from the touch window remains in the document and is no longer
/// part of the range replaced by later candidate changes or Backspace.
public struct TouchSmartHostTextState {
    public struct Edit: Equatable {
        public let deleteCount: Int
        public let insertion: String

        public var isEmpty: Bool { deleteCount == 0 && insertion.isEmpty }
    }

    public private(set) var editableText = ""

    public init() {}

    public mutating func update(to newText: String, evictedPrefix: String = "") -> Edit {
        let oldText: String
        let desiredText: String
        if !evictedPrefix.isEmpty, editableText.hasPrefix(evictedPrefix) {
            oldText = String(editableText.dropFirst(evictedPrefix.count))
            desiredText = newText
        } else {
            oldText = editableText
            desiredText = evictedPrefix + newText
        }
        let edit = Self.difference(from: oldText, to: desiredText)
        editableText = newText
        return edit
    }

    public mutating func finish(with text: String) -> Edit {
        let edit = Self.difference(from: editableText, to: text)
        editableText = ""
        return edit
    }

    public mutating func cancel() -> Edit {
        finish(with: "")
    }

    public mutating func reset() {
        editableText = ""
    }

    private static func difference(from oldText: String, to newText: String) -> Edit {
        let shared = zip(oldText, newText).prefix { $0 == $1 }.count
        return Edit(
            deleteCount: oldText.count - shared,
            insertion: String(newText.dropFirst(shared))
        )
    }
}

File: Source/Loaders/Android-IME/app/src/main/java/tw/chichi77/keykey/android/TouchSmartHostTextState.java
package tw.chichi77.keykey.android;

/** The committed host-text suffix that touch Smart Mandarin may still change. */
final class TouchSmartHostTextState {
    record Edit(int deleteCount, String insertion) {
        boolean isEmpty() { return deleteCount == 0 && insertion.isEmpty(); }
    }

    private String editableText = "";

    String editableText() { return editableText; }

    Edit update(String newText, String evictedPrefix) {
        String oldText;
        String desiredText;
        if (!evictedPrefix.isEmpty() && editableText.startsWith(evictedPrefix)) {
            oldText = editableText.substring(evictedPrefix.length());
            desiredText = newText;
        } else {
            oldText = editableText;
            desiredText = evictedPrefix + newText;
        }
        Edit edit = difference(oldText, desiredText);
        editableText = newText;
        return edit;
    }

    Edit finish(String text) {
        Edit edit = difference(editableText, text);
        editableText = "";
        return edit;
    }

    Edit cancel() { return finish(""); }

    void reset() { editableText = ""; }

    private static Edit difference(String oldText, String newText) {
        int oldEnd = 0;
        int newEnd = 0;
        while (oldEnd < oldText.length() && newEnd < newText.length()) {
            int oldCodePoint = oldText.codePointAt(oldEnd);
            int newCodePoint = newText.codePointAt(newEnd);
            if (oldCodePoint != newCodePoint) break;
            oldEnd += Character.charCount(oldCodePoint);
            newEnd += Character.charCount(newCodePoint);
        }
        return new Edit(oldText.length() - oldEnd, newText.substring(newEnd));
    }
}
